components/cloud/azure/src/kms.rs

```diff
@@ -0,0 +1,337 @@
+};
+use tikv_util::box_err;
+
+use crate::{ClientCertificateCredentialExt, STORAGE_VENDOR_NAME_AZURE};
+
+/// Use 256 bits for data key as default.
+const DEFAULT_DATAKEY_SIZE: u8 = 32;
+const ENCRYPTION_VENDOR_NAME_AZURE_KMS: &str = STORAGE_VENDOR_NAME_AZURE;
+
+pub struct AzureKms {
+    tenant_id: String,
+    client_id: String,
+    /// Keyvault client to encrypt/decrypt data key.
+    client: KeyClient,
+    current_key_id: KeyId,
+    keyvault_url: String,
+    /// Hsm client to get random bytes for generating data key.
+    hsm_client: KeyClient,
+    hsm_name: String,
+    hsm_url: String,
+}
+
+impl std::fmt::Debug for AzureKms {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let keyvault_client = AzureKeyVaultClientDebug {
+            tenant_id: self.tenant_id.clone(),
+            client_id: self.client_id.clone(),
+            keyvault_url: self.keyvault_url.clone(),
+        };
+        let hsm_client = AzureHsmClientDebug {
+            hsm_name: self.hsm_name.clone(),
+            hsm_url: self.hsm_url.clone(),
+        };
+        f.debug_struct("AzureKms")
+            .field("keyvault_client", &keyvault_client)
+            .field("current_key_id", &self.current_key_id)
+            .field("hsm_client", &hsm_client)
+            .finish()
+    }
+}
+
+impl AzureKms {
+    #[inline]
+    fn new_with_credentials<Creds>(
+        config: Config,
+        keyvault_credentials: Creds,
+        hsm_credentials: Creds,
+    ) -> Result<Self>
+    where
+        Creds: TokenCredential + Send + Sync + 'static,
+    {
+        assert!(config.azure.is_some());
+        let azure_cfg = config.azure.unwrap();
+        let keyvault_client = new_key_client(&azure_cfg.keyvault_url, keyvault_credentials)?;
+        let hsm_client = new_key_client(&azure_cfg.hsm_url, hsm_credentials)?;
+        Ok(Self {
+            client: keyvault_client,
+            current_key_id: config.key_id,
+            tenant_id: azure_cfg.tenant_id,
+            client_id: azure_cfg.client_id,
+            keyvault_url: azure_cfg.keyvault_url,
+            hsm_client,
+            hsm_name: azure_cfg.hsm_name,
+            hsm_url: azure_cfg.hsm_url,
+        })
+    }
+
+    pub fn new(config: Config) -> Result<Self> {
+        assert!(config.azure.is_some());
+        let azure_cfg = config.azure.clone().unwrap();
+        // Priority: explicit certificate > path of local certificate > client secret.
+        // And credentials for accessing KeyVault and Hsm should be different.
+        if let Some(certificate) = azure_cfg.client_certificate.clone() {
+            // Certificate to accessing KeyVault.
+            let (keyvault_credential, hsm_credential) = (
+                ClientCertificateCredentialExt::new(
+                    azure_cfg.tenant_id.clone(),
+                    azure_cfg.client_id.clone(),
+                    certificate.clone(),
+                    "".to_owned(),
```

Comment:
Empty `certificate_pass`?  What if it's protected by a password?

Reply:
Got. I'll expose this configuration, I've missed this point.